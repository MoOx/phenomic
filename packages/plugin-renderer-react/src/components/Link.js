// @flow
import React from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import { Link as RouterLink } from "react-router";

const origin = url =>
  typeof url === "object" &&
  // jsdom can return "null" string...
  ((url.origin !== "null" && url.origin) ||
    // // IE does not correctly handle origin, maybe Edge does...
    url.protocol + "//" + url.hostname + (url.port ? ":" + url.port : ""));

type PropsType = {
  className?: string,
  activeClassName?: string,
  to?: string,
  href?: string,
  children?: any
};

function Link(props: PropsType, { router }: { router: Object }) {
  // eslint-disable-next-line no-unused-vars
  const { to, href, ...otherProps } = props;
  const url = props.to || props.href || "";

  const simpleLink = <a {...otherProps} href={url} />;

  // static rendering
  if (typeof document === "undefined") {
    return simpleLink;
  }

  const toLink = document.createElement("a");
  toLink.href = url;

  // @todo handle properly use base only
  // in v0.x we were relying on the base url
  // if (
  //   // parent absolute url with the same domain
  //   // should not be Link
  //   url.indexOf("://") > -1 //&&
  //   url.indexOf(process.env.PHENOMIC_USER_URL) === -1
  // ) {
  //   return simpleLink;
  // }

  if (
    // jsdom have url data directly into "url" property, so just in case
    // it's like that on some browsers...
    origin(toLink) === origin(window.location)
    // we might want to restrict Link to path including the pathname
    // but this will require to preprend pathname to all Links from the
    // collection, which sucks.
    // If people wants to use Link for a same domain, but in the parent path,
    // you will need to includes the entire url, / won't work at it will use
    // the react-router basename defined by Phenomic.
    // &&
    // toLink.pathname.includes(process.env.PHENOMIC_USER_PATHNAME)
    // toLink.pathname.indexOf(process.env.PHENOMIC_USER_PATHNAME) > -1
  ) {
    const classNames = cx(props.className, {
      [props.activeClassName || ""]:
        router &&
          (router.isActive({ pathname: url }) ||
            router.isActive({ pathname: url + "index.html" })) &&
          Boolean(props.activeClassName)
    });
    return (
      <RouterLink
        {...otherProps}
        // react-router seems to be pretty dumb with relative links...
        // this way we have a proper link
        to={toLink.href}
        {...(classNames ? { classNames } : {})}
      />
    );
  }

  return simpleLink;
}

Link.propTypes = {
  children: PropTypes.node,
  to: PropTypes.string,
  href: PropTypes.string,
  className: PropTypes.string,
  activeClassName: PropTypes.string
};

Link.contextTypes = {
  router: PropTypes.object.isRequired
};

Link.displayName = "Link";

export default Link;
