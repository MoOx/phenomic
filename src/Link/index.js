// @flow
import React, { PropTypes } from "react"
import cx from "classnames"

import { Link as RouterLink } from "react-router"

export function Link(
  props: {
    className: string,
    activeClassName: string,
    to: string,
    children: any,
  },
  { router }: { router: Object }
): React$Element {
  return (
    <RouterLink
      { ...props }
      className={ cx(props.className, {
        [props.activeClassName]: (router && (
          router.isActive({ pathname: props.to }) ||
          router.isActive({ pathname: props.to + "index.html" })
        )) && props.activeClassName,
      }) }
    >
      { props.children }
    </RouterLink>
  )
}

Link.propTypes = {
  children: PropTypes.node,
  to: PropTypes.string.isRequired,
  className: PropTypes.string,
  activeClassName: PropTypes.string,
}

Link.contextTypes = {
  router: PropTypes.object.isRequired,
}

Link.displayName = "Link"

/*
  exported as default and Link so people can easily switch their
  import { Link } from "react-router"
  to
  import { Link } from "phenomic/lib/Link"
  or
  import Link from "phenomic/lib/Link"
*/
export default Link
