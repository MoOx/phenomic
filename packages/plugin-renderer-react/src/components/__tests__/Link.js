/* eslint-disable react/no-multi-comp */

import React from "react";
import { renderJSX } from "jsx-test-helpers";

import Link from "../Link";

test("Link works with ``to``", () => {
  expect(renderJSX(<Link to="/b">{"c"}</Link>)).toMatchSnapshot();
});

test("Link works with ``href``", () => {
  expect(renderJSX(<Link href="/b">{"c"}</Link>)).toMatchSnapshot();
});

test("Link works with external ``href``", () => {
  expect(renderJSX(<Link href="http://b.com">{"c"}</Link>)).toMatchSnapshot();
});

test("Link works with external ``href``", () => {
  expect(renderJSX(<Link href="http://b.com">{"c"}</Link>)).toMatchSnapshot();
});

test("Link works with internal ``href``", () => {
  expect(
    renderJSX(<Link href="test://url.tld/internal">{"c"}</Link>)
  ).toMatchSnapshot();
  expect(renderJSX(<Link href="/internal">{"c"}</Link>)).toMatchSnapshot();
});

test("Link activeClassName?", () => {
  expect(
    renderJSX(<Link href="/internal" activeClassName="activeCx">{"c"}</Link>, {
      router: { isActive: () => true }
    })
  ).toMatchSnapshot();
  expect(
    renderJSX(<Link href="/internal" activeClassName="activeCx">{"c"}</Link>, {
      router: { isActive: () => false }
    })
  ).toMatchSnapshot();
});
