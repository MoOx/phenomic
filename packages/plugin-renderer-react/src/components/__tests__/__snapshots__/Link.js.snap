// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link activeClassName? 1`] = `
"<Link
  activeClassName=\\"activeCx\\"
  classNames=\\"activeCx\\"
  style={{}}
  to=\\"test://url.tld/internal\\"
>
  c
</Link>"
`;

exports[`Link activeClassName? 2`] = `
"<Link
  activeClassName=\\"activeCx\\"
  style={{}}
  to=\\"test://url.tld/internal\\"
>
  c
</Link>"
`;

exports[`Link works with \`\`href\`\` 1`] = `
"<Link
  style={{}}
  to=\\"test://url.tld/b\\"
>
  c
</Link>"
`;

exports[`Link works with \`\`to\`\` 1`] = `
"<Link
  style={{}}
  to=\\"test://url.tld/b\\"
>
  c
</Link>"
`;

exports[`Link works with external \`\`href\`\` 1`] = `
"<a href=\\"http://b.com\\">
  c
</a>"
`;

exports[`Link works with external \`\`href\`\` 2`] = `
"<a href=\\"http://b.com\\">
  c
</a>"
`;

exports[`Link works with internal \`\`href\`\` 1`] = `
"<Link
  style={{}}
  to=\\"test://url.tld/internal\\"
>
  c
</Link>"
`;

exports[`Link works with internal \`\`href\`\` 2`] = `
"<Link
  style={{}}
  to=\\"test://url.tld/internal\\"
>
  c
</Link>"
`;
